homeassistant/components/elmo_alarm/__init__.py

```diff
@@ -0,0 +1,265 @@
+
+    hass.async_create_task(
+        async_load_platform(hass, "alarm_control_panel", DOMAIN, {}, config)
+    )
+
+    async def update():
+        _LOGGER.debug("Connecting to e-connect to retrieve states")
+        await client.update()
+        async_dispatcher_send(hass, SIGNAL_ARMING_STATE_CHANGED, client.state)
+
+        for zone in client.zones:
+            async_dispatcher_send(hass, SIGNAL_ZONE_CHANGED, zone)
+
+        for inp in client.inputs:
+            async_dispatcher_send(hass, SIGNAL_INPUT_CHANGED, inp)
+
+        async_call_later(
+            hass,
+            scan_interval.total_seconds(),
+            lambda _: hass.async_create_task(update()),
+        )
+
+    await update()
+
+    return True
+
+
+class ElmoClientWrapper(ElmoClient):
+    """Wrapping the Elmo client class to adapt it to home assistant."""
+
+    def __init__(self, host, vendor, username, password, states_config):
+        """Initialize the elmo client wrapper."""
+
+        self._username = username
+        self._password = password
+        self._states_config = states_config
+        self._data = None
+        self.state = None
+        self.states = None
+        self.zones = None
+        self.inputs = None
+
+        ElmoClient.__init__(self, host, vendor)
+
+    async def update(self):
+        """Get updates and refresh internal states."""
+        try:
+            data = self.check()
+        except PermissionDenied as exception:
+            _LOGGER.warning("Invalid session, trying to authenticate %s", exception)
+            try:
+                self.auth(self._username, self._password)
+                data = self.check()
+            except PermissionDenied as exception:
+                _LOGGER.warning("Invalid credentials: %s", exception)
+            except HTTPError as exception:
+                _LOGGER.warning(
+                    "Got HTTP error when authenticating: %s", exception,
+                )
+
+        if self._data is None:
+            await self._configure_states()
+        self._data = data
+        await self._update_arm_state()
+        await self._update_zone_state()
+        await self._update_input_state()
+
+    async def _configure_states(self):
+        """Initialize the elmo alarm states."""
+        state_config = {}
+        for state in self._states_config:
+            state_config[state[CONF_NAME]] = state[CONF_ZONES]
+        self.states = state_config
+
+    async def _update_arm_state(self):
+        """Update the elmo alarm states."""
+        areas_armed = self._data["areas_armed"]
+
+        if not areas_armed:
+            self.state = STATE_ALARM_DISARMED
+        else:
+            armed_indexes = [area_data["index"] + 1 for area_data in areas_armed]
+
+            [state] = [
+                k
+                for k, v in filter(
+                    lambda state: set(state[1]) == set(armed_indexes),
+                    self.states.items(),
+                )
+            ] or [None]
```

Comment:
Please make this code block more readable by using self-explanatory variable names, maybe filter first and then do the loop and add a comment explaining what this code is doing. 
Tbh I have currently trouble figuring that out 😅

Reply:
I agree, I tried to make it really pythonic! :)

Anyway a reviewed it and now it should be really easy to understand.